### PR TITLE
First attempt at converting input with wana_kana

### PR DESCRIPTION
**Pull request by ian-h-chamberlain**
Relates to #4 (but maybe doesn't close it).

Based on the discussion we had I tried to implement "Option 3" – so far, it seems to work _pretty_ well, there might be a few edge cases I haven't found (and there's one which seems like a `wana_kana` bug, ソフトウェア -> `softouea`)

I took some liberties with the word list format, but it can probably be changed fairly easily. My initial thought was to use kana in the game data itself, and convert user input to match kana – but I think keeping the format as it was and converting back to romaji might be simpler for e.g. what happens if the user hits backspace?

Let me know what you think of this approach. For now I will leave this as a draft PR just to get the conversation started.

P.S. I also made a couple build/dependency changes to get the project building in debug mode on my machine, but I can definitely back these out if you're concerned about them.

### CODE REVIEW

**Comment by rparrett** [comment]
Thanks for the PR! The PR itself is really well-done and easy to follow, so double-thanks for that.

This deserves a bit more attention/thought than I can give this weekend (currently in the middle of a move). I haven't even run the code yet, but I'll put some uninformed thoughts here for now and try to revisit in a bit.

I think that I generally like this approach. It's not quite what I thought I was describing in option 3, but it seems better.

I am not completely sold on the changes to the data files. I think that the direction I'd like to head with this is towards having simple flat files with one-word-per-line. I'd like to make it super easy for normal people to create word lists in the future.

Although I do think you improved the names of the `WordListKind`s.

I would like to eventually create separate loaders for e.g.

`english.plain.txt` (one word per line, input/rendered is the same)
`n5kanji.jp.txt` (one word per line, parens for kanji/numerals, kana automatically handled)
`whatever.parens.txt` (one word per line, `in(ren)put(dered)`)

The way things currently work is a consequence of not being able to register separate loaders for `.plain.txt`, and `.jp.txt`, but that is now possible with bevy 0.5.

File: assets/data/game.ron [hunk 1]
```diff
@@ -253,7 +254,7 @@ GameData(
         "crab",
         "criminal",
         "crooked",
-        "cunning",
```
Comment (rparrett) [inline assets/data/game.ron]: slightly over-zealous search/replace

**Comment by rparrett** [comment]
Can the「ソフトウェア」situation be worked around with a `custom_kana_mapping`? edit: This doesn't seem possible in the rust crate.

edit: I thought this could be due to implicit "hiragananess" from using our lower-case input with `to_kana`, but it does seem like it might just be a bug.

「つづく」 is also malfunctioning, but I think that's just mistake in the normalized text in the word list. (should be `tsuzuku`)

File: assets/data/game.ron (hunk 1 above)
Comment (ian-h-chamberlain) [reply assets/data/game.ron]: hehe, yep. I noticed `Annotated` -> `Anotated` but forgot to check the english words.

**Comment by ian-h-chamberlain** [comment]
> I would like to eventually create separate loaders for e.g.
> 
> `english.plain.txt` (one word per line, input/rendered is the same)
> `n5kanji.jp.txt` (one word per line, parens for kanji/numerals, kana automatically handled)
> `whatever.parens.txt` (one word per line, `in(ren)put(dered)`)
> 
> The way things currently work is a consequence of not being able to register separate loaders for `.plain.txt`, and `.jp.txt`, but that is now possible with bevy 0.5.

Ah ok, I guess this is why you had the inputs as plain strings, that makes more sense now. I can certainly undo the changes I had there since it seems easier for people to contribute words.

> Can the「ソフトウェア」situation be worked around with a custom_kana_mapping? edit: This doesn't seem possible in the rust crate.
> 
> edit: I thought this could be due to implicit "hiragananess" from using our lower-case input with to_kana, but it does seem like it might just be a bug.
> 
> 「つづく」 is also malfunctioning, but I think that's just mistake in the normalized text in the word list. (should be tsuzuku)

Apparently, this is a bug even in the upstream JS library:
Screen Shot 2021-05-15 at 4 55 40 PM

However, I think you're right that the `custom_kana_mapping` feature, if ported to the rust version, would probably resolve it.

As for 「つづく」I think `tsuduku` is the correct romaji for that? But the `to_romaji` conversion seems to output `tsuzuku`, which would convert to 「つずく」. So this may be another case of the library having some inconsistencies...

One thing I considered, rather than converting the user input `romaji -> kana -> romaji` and comparing to the word list, was converting the user input `romaji -> kana` and the word list `romaji -> kana` and comparing the kana. Alternately, the word list could be provided in kana already. I wasn't sure if this would mess up the render highlighting or backspace behavior, etc. so I opted to do this round trip to romaji instead, but I wonder if it would be more reliable that way.

There is also the `imemode: true`, which maybe only makes sense to use on the user input, and not when converting the word list? I had some issues with converting things like `sennsei` (it would pass through something like 「せんnせい」), which is how I arrived and removed all the `nn` from the word list.

**Comment by ian-h-chamberlain** [comment]
Also, no need to rush or anything! I will keep experimenting but certainly this is not urgent, and I don't mind waiting for feedback if you have life stuff to deal with. :+1:

**Comment by rparrett** [comment]
> However, I think you're right that the `custom_kana_mapping` feature, if ported to the rust version, would probably resolve it.

It doesn't seem like this would be super difficult to add, if it turns out that wana_kana is behaving in an expected way.

> As for 「つづく」I think `tsuduku` is the correct romaji for that? But the `to_romaji` conversion seems to output `tsuzuku`, which would convert to 「つずく」. So this may be another case of the library having some inconsistencies...

Yep, oops.

> One thing I considered, rather than converting the user input `romaji -> kana -> romaji` and comparing to the word list, was converting the user input `romaji -> kana` and the word list `romaji -> kana` and comparing the kana. Alternately, the word list could be provided in kana already. I wasn't sure if this would mess up the render highlighting or backspace behavior, etc. so I opted to do this round trip to romaji instead, but I wonder if it would be more reliable that way.

I was surprised that the full round trip works as well as it does. How would you propose handling kanji, numerals, etc in the `romaji -> kana<=>kana <- romaji` scenario?

I am willing to make some small sacrifices in the highlighting. e.g. `シ(sh)ャ(a)ツ(tsu)` -> `シャ(sha)ツ(tsu)` is fine.

> There is also the `imemode: true`, which maybe only makes sense to use on the user input, and not when converting the word list? I had some issues with converting things like `sennsei` (it would pass through something like 「せんnせい」), which is how I arrived and removed all the `nn` from the word list.

I think it's important that input while playing be flexible, but I am not as concerned about the word lists, because I hope to have automatic kana handling there anyhow. I'd like `小(chii)さ(sa)い(i)` to be written as `小(chii)さい`.

**Comment by ian-h-chamberlain** [comment]
Hey sorry been pretty busy recently, but it looks like you've done some work implementing this feature in the meantime.

It seems like you're leaning away from `wana_kana` which is fine, but I did want to mention this

> I think it's important that input while playing be flexible, but I am not as concerned about the word lists, because I hope to have automatic kana handling there anyhow. I'd like 小(chii)さ(sa)い(i) to be written as 小(chii)さい.

I think having something like `小(ちい)さい` might be more flexible, since you can just convert the user input to kana and match the chunks that way, rather than relying on multiple possibilities with romaji (like `{"ti", "chi"}` `{"n" + consonant, "nn"}`, etc.). It looks like you started that process in #6 and I think it makes sense to continue.

I guess for katakana words you would still need to either have something like `コ(こ)ン(ん)ピュ(ぴゅ)ータ(た)ー` (ugh, not pretty) _or_ do the conversion to both hiragana + katakana and allow it to match either (seems more powerful).

Based on that PR, is there anything you want me to keep here, or should I just close this in favor of your changes?

**Comment by ian-h-chamberlain** [comment]
Oh, one more thing, if you wanted to simplify the parsing code a bit you could probably use a regex like [this playground example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=d18a5fd2dba8f7d19b4b95a9c068081f).

As written it only does one character, but might be a little simpler than a custom parser, depending on what you need it to do. This is basically what I did in python when I changed the data format a little for this PR.

**Comment by rparrett** [comment]
> Based on that PR, is there anything you want me to keep here, or should I just close this in favor of your changes?

I actually am leaning more towards your general approach. I wasn't initially a fan of individual typing targets being language-specific, but it definitely feels cleaner than what I ended up doing to them. I think that #6 could eventually work, but this PR feels like a better starting point.

If you're still feeling that way and still motivated to experiment with it, please do so. I'm not sure when I'll get another big chunk of time to play around with it.

> I guess for katakana words you would still need to either have something like `コ(こ)ン(ん)ピュ(ぴゅ)ータ(た)ー` (ugh, not pretty) _or_ do the conversion to both hiragana + katakana and allow it to match either (seems more powerful).

Yeah, that's not ideal. I think that `input -> hiragana <=> hiragana <- hiragana / katakana <- wordlist` also seems potentially messy to me.

`input -> normalized romaji <=> normalized romaji <- wordlist` feels the cleanest to me, but it doesn't seem like the `wana_kana` `romaji -> kana -> romaji` roundtrip is going to work out for that.

I am wondering if the amount of flexibility we actually need to make most people happy is possible with just some simple regex based processing on the input or something.

> I think having something like `小(ちい)さい` might be more flexible

Agreed. I think that's how I'd like the Japanese word lists to generally look.

> but might be a little simpler than a custom parser, depending on what you need it to do. This is basically what I did in python when I changed the data format a little for this PR.

Yeah, the continued use of `nom` is purely for learning / funsies and I'd definitely accept a simpler solution that still works with the desired wordlist format.

**Pull request closed without merging**